pkg/server/grpc/server.go

```diff
@@ -88,6 +90,60 @@ func (s *Server) Run() error {
 }
 
 // New creates a new instance of `IstioCAServiceServer`.
-func New(ca ca.CertificateAuthority, port int) *Server {
-	return &Server{ca, port}
+func New(ca ca.CertificateAuthority, hostname string, port int) *Server {
+	return &Server{
+		ca:       ca,
+		hostname: hostname,
+		port:     port,
+	}
+}
+
+func (s *Server) createTLSServerOption() grpc.ServerOption {
+	config := &tls.Config{
+		GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+			if s.certificate == nil || !isValid(s.certificate) {
+				// Apply new certificate if there isn't one yet, or the one has become invalid.
+				newCert, err := s.applyServerCertificate()
+				if err != nil {
+					return nil, err
+				}
+				s.certificate = newCert
+			}
+			return s.certificate, nil
+		},
+	}
+	return grpc.Creds(credentials.NewTLS(config))
+}
+
+func (s *Server) applyServerCertificate() (*tls.Certificate, error) {
+	opts := ca.CertOptions{
+		Host:       s.hostname,
+		RSAKeySize: 2048,
+	}
+
+	csrPEM, privPEM, err := ca.GenCSR(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	certPEM, err := s.ca.Sign(csrPEM)
+	if err != nil {
+		return nil, err
+	}
+
+	cert, err := tls.X509KeyPair(certPEM, privPEM)
+	if err != nil {
+		return nil, err
+	}
+	return &cert, nil
+}
+
+func isValid(cert *tls.Certificate) bool {
+	leaf := cert.Leaf
+	if leaf == nil {
+		return false
+	}
+	now := time.Now()
+	return leaf.NotBefore.Add(certValidationBuffer).Before(now) &&
```

Comment:
Do we need the buffer for `NotBefore`? Won't this cause a certificate to be considered invalid right after issuance?